pkgs/development/compilers/rust/rustc.nix

```diff
@@ -136,6 +136,9 @@ in stdenv.mkDerivation rec {
 
     # Useful debugging parameter
     # export VERBOSE=1
+
+    substituteInPlace compiler/rustc_target/src/spec/linux_musl_base.rs \
```

Comment:
Could make this conditional to `stdenv.hostPlatform.isMusl` too, since otherwise it changes the behavior for people using the pkgsStatic.buildPackages.rustc manually. They would now have to manually set `+crt-static` to get a static binary, which they might not expect since upstream is explicitly keeping the default.

Reply:
Actually I don't think it's necessary. Most people who do this kind of stuff manually will use rustup.

Reply:
I think it's a good idea not to change the upstream behaviour where a user hasn't explicitly requested it, so I'm going to make this change.  But I think it should be `stdenv.targetPlatform` so it's correct when building a musl cross compiler.

Reply:
> But I think it should be stdenv.targetPlatform so it's correct when building a musl cross compiler.

Not sure about this. When the compiler is a cross-compiler from glibc to musl, the build.rs scripts will also run on glibc and this setting does not matter.
If I remember correctly what hostPlatform and targetPlatform are...

Reply:
> Not sure about this. When the compiler is a cross-compiler from glibc to musl, the build.rs scripts will also run on glibc and this setting does not matter.
> If I remember correctly what hostPlatform and targetPlatform are...

But this is the change for the default value of crt-static for all Musl builds in rustc, not the one for build.rs scripts in Cargo.

> For example, targetPlatform.isMusl would be true for pkgsStatic.buildPackages.rustc, even though the change is not needed there?

Yes, the actual check I would do here is `stdenv.targetPlatform.isMusl && !stdenv.targetPlatform.isStatic`.



> Or the opposite, if we are building a cross-compiler from musl to glibc, targetPlatform.isMusl will be false, but it _should_ still apply the change because the build.rs will be run with musl.

Not sure about this one.  I'll have to test it.

Reply:
> > Or the opposite, if we are building a cross-compiler from musl to glibc, targetPlatform.isMusl will be false, but it _should_ still apply the change because the build.rs will be run with musl.
> 
> Not sure about this one. I'll have to test it.

I tested this by building `pkgsMusl.pkgsCross.aarch64-multiplatform.{fd,cargo}`, and both built fine even without this change.  I don't entirely understand why, but it seems like just enabling it when `targetPlatform` is non-static Musl works fine.